{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Immunity loss simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from opqua.model import Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "cannot unpack non-iterable NoneType object",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-d84cae6a93e3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlinkPopulationsHostHostContact\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'my_population_B'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'my_population_A'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2e-1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 73\u001b[0;31m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1000\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtime_sampling\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     74\u001b[0m data = model.saveToDataFrame(\n\u001b[1;32m     75\u001b[0m     \u001b[0;34m'TESTS_immunity_loss/immunity_loss_test.csv'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/opqua_immunity/opqua/model.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, t0, tf, time_sampling, host_sampling, vector_sampling)\u001b[0m\n\u001b[1;32m    764\u001b[0m         \u001b[0msim\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mGillespie\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    765\u001b[0m         self.history = sim.run(\n\u001b[0;32m--> 766\u001b[0;31m             \u001b[0mt0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtime_sampling\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhost_sampling\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvector_sampling\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    767\u001b[0m             )\n\u001b[1;32m    768\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/opqua_immunity/opqua/internal/gillespie.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, t0, tf, time_sampling, host_sampling, vector_sampling, print_every_n_events)\u001b[0m\n\u001b[1;32m    666\u001b[0m                                     e, self.model.populations[\n\u001b[1;32m    667\u001b[0m                                         \u001b[0mpopulation_ids\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 668\u001b[0;31m                                         ], ( u - r_cum + r[e,p] ) / r[e,p]\n\u001b[0m\u001b[1;32m    669\u001b[0m                                     ) # do corresponding action,\n\u001b[1;32m    670\u001b[0m                                       \u001b[0;31m# feed in renormalized random number\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/opqua_immunity/opqua/internal/gillespie.py\u001b[0m in \u001b[0;36mdoAction\u001b[0;34m(self, act, pop, rand)\u001b[0m\n\u001b[1;32m    544\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    545\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mact\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDEIMMUNIZE_HOST\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 546\u001b[0;31m             \u001b[0mpop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdeimmunizeHost\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    547\u001b[0m             \u001b[0mchanged\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    548\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/opqua_immunity/opqua/internal/population.py\u001b[0m in \u001b[0;36mdeimmunizeHost\u001b[0;34m(self, rand)\u001b[0m\n\u001b[1;32m   1299\u001b[0m         \u001b[0mhost\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhosts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex_host\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1300\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1301\u001b[0;31m         \u001b[0mhost\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdeimmunize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1302\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdateIndividualCoefficients\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhost\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1303\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/opqua_immunity/opqua/internal/host.py\u001b[0m in \u001b[0;36mdeimmunize\u001b[0;34m(self, rand)\u001b[0m\n\u001b[1;32m    405\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimmunity_sequences\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    406\u001b[0m             ]\n\u001b[0;32m--> 407\u001b[0;31m         \u001b[0mindex_genome\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mrand\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetWeightedRandomGenome\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mrand\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mweights\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    408\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    409\u001b[0m         \u001b[0mdeimmunize_genome\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimmunity_sequences\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex_genome\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: cannot unpack non-iterable NoneType object"
     ]
    }
   ],
   "source": [
    "def myImmuneWeights(genome,immune_seq):\n",
    "    return Model.perfectMatchImmunity(\n",
    "        genome, immune_seq, weight=1\n",
    "        )\n",
    "        # Perfect matching: any mismatches between pathogen genome and immune\n",
    "        # memory generate no immunity. Only perfect matches generate 100%\n",
    "        # immunity.\n",
    "\n",
    "model = Model()\n",
    "model.newSetup( # Now, we'll define our new setup:\n",
    "    'my_setup_A', preset='host-host', # Use default host-host parameters.\n",
    "    possible_alleles='AB',\n",
    "        # Define \"letters\" in the \"genome\", or possible alleles for each locus.\n",
    "        # Each locus can have different possible alleles if you define this\n",
    "        # argument as a list of strings, but here, we take the simplest\n",
    "        # approach.\n",
    "    num_loci=3,\n",
    "        # Define length of \"genome\", or total number of alleles.\n",
    "    mutate_in_host=1e-2,\n",
    "        # Modify de novo mutation rate of pathogens when in host to get some\n",
    "        # evolution!\n",
    "    immunity_acquisition_rate_host=1e-2,\n",
    "        # rate at which immunity is acquired within infected individuals\n",
    "    immunity_loss_rate_host=1e-2,\n",
    "        # rate at which immunity is lost within infected individuals\n",
    "    immunityWeightsHost=myImmuneWeights,\n",
    "        # immunity function that evaluates effect of immune memory on pathogen\n",
    "        # genomes.\n",
    "    contact_rate_host_host=2e-1,\n",
    "    recovery_rate_host=1e-1,\n",
    "    )\n",
    "\n",
    "model.newSetup( # Now, we'll define our new setup:\n",
    "    'my_setup_B', preset='host-host', # Use default host-host parameters.\n",
    "    possible_alleles='AB',\n",
    "        # Define \"letters\" in the \"genome\", or possible alleles for each locus.\n",
    "        # Each locus can have different possible alleles if you define this\n",
    "        # argument as a list of strings, but here, we take the simplest\n",
    "        # approach.\n",
    "    num_loci=3,\n",
    "        # Define length of \"genome\", or total number of alleles.\n",
    "    mutate_in_host=1e-2,\n",
    "        # Modify de novo mutation rate of pathogens when in host to get some\n",
    "        # evolution!\n",
    "    immunity_acquisition_rate_host=1e-2,\n",
    "        # rate at which immunity is acquired within infected individuals\n",
    "    immunity_loss_rate_host=1e-2,\n",
    "        # rate at which immunity is lost within infected individuals\n",
    "    immunityWeightsHost=myImmuneWeights,\n",
    "        # immunity function that evaluates effect of immune memory on pathogen\n",
    "        # genomes.\n",
    "    contact_rate_host_host=2e-1,\n",
    "    recovery_rate_host=1e-2,\n",
    "    )\n",
    "\n",
    "model.newPopulation('my_population_A','my_setup_A', num_hosts=int(50))\n",
    "model.newPopulation('my_population_B','my_setup_B', num_hosts=int(50))\n",
    "\n",
    "model.addPathogensToHosts( 'my_population_A',{'AAA':10} )\n",
    "    # We will start off the simulation with a suboptimal pathogen genome,\n",
    "    # \"AAA\". Throughout the course of the simulation, we should see this genome\n",
    "    # be outcompeted by other pathogen genotypes as the host population acquires\n",
    "    # resistance to each genome.\n",
    "model.addPathogensToHosts( 'my_population_B',{'AAA':10} )\n",
    "    # We will start off the simulation with a suboptimal pathogen genome,\n",
    "    # \"AAA\". Throughout the course of the simulation, we should see this genome\n",
    "    # be outcompeted by other pathogen genotypes as the host population acquires\n",
    "    # resistance to each genome.\n",
    "\n",
    "model.linkPopulationsHostHostContact('my_population_A','my_population_B',2e-1)\n",
    "model.linkPopulationsHostHostContact('my_population_B','my_population_A',2e-1)\n",
    "\n",
    "model.run(0,1000,time_sampling=0)\n",
    "data = model.saveToDataFrame(\n",
    "    'TESTS_immunity_loss/immunity_loss_test.csv'\n",
    "    )\n",
    "\n",
    "graph_populations = model.populationsPlot( # Plot infected hosts per population over time.\n",
    "    'TESTS_immunity_loss/populations_plot.png', \n",
    "    data,\n",
    "    num_top_populations=2, # plot all 2 populations\n",
    "    y_label='Infected hosts' # change y label\n",
    "    )\n",
    "\n",
    "graph_composition = model.compositionPlot(\n",
    "        # Create a plot to track pathogen genotypes across time.\n",
    "    'TESTS_immunity_loss/pathogen_composition.png',\n",
    "    data,\n",
    "    type_of_composition='Pathogens',\n",
    "    num_top_sequences=8,\n",
    "        # Track the 8 most represented genomes overall (only 8 possible).\n",
    "    )\n",
    "\n",
    "graph_composition = model.compositionPlot(\n",
    "        # Create a plot of genotypes in the hosts' immune memories across time.\n",
    "    'TESTS_immunity_loss/immunity_composition.png', \n",
    "    data,\n",
    "    type_of_composition='Immunity', y_label='Genomes in immune memory',\n",
    "    num_top_sequences=8,\n",
    "        # Track the 8 most represented genomes overall (only 8 possible).\n",
    "    )\n",
    "\n",
    "graph_compartments = model.compartmentPlot(\n",
    "    'TESTS_immunity_loss/reassortment_compartments.png', \n",
    "    data\n",
    "    )\n",
    "    # Also generate a normal compartment plot. Notice the total number of\n",
    "    # infections in the composition plot can exceed the number of infected hosts\n",
    "    # in the compartment plot. This happens because a single host infected by\n",
    "    # multiple genotypes is counted twice in the former, but not the latter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "d203a7fbe37afbb990fedfc21c321928443618f3d7b991e0237ff71906aa031f"
  },
  "kernelspec": {
   "display_name": "Python 3.7.0 ('datascience')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
